<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0220ENSkillsNetwork900-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


<h1>Extracting and Visualizing Stock Data</h1>
<h2>Description</h2>


Extracting essential data from a dataset and displaying it is a necessary part of data science; therefore individuals can make correct decisions based on the data. In this assignment, you will extract some stock data, you will then display this data in a graph.


<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li>Define a Function that Makes a Graph</li>
        <li>Question 1: Use yfinance to Extract Stock Data</li>
        <li>Question 2: Use Webscraping to Extract Tesla Revenue Data</li>
        <li>Question 3: Use yfinance to Extract Stock Data</li>
        <li>Question 4: Use Webscraping to Extract GME Revenue Data</li>
        <li>Question 5: Plot Tesla Stock Graph</li>
        <li>Question 6: Plot GameStop Stock Graph</li>
    </ul>
<p>
    Estimated Time Needed: <strong>30 min</strong></p>
</div>

<hr>


***Note***:- If you are working in IBM Cloud Watson Studio, please replace the command for installing nbformat from `!pip install nbformat==4.2.0` to simply `!pip install nbformat`


In [ ]:
!pip install yfinance==0.1.67
!mamba install bs4==4.10.0 -y
!pip install nbformat==4.2.0

In [ ]:
import yfinance as yf
import pandas as pd
import requests
from bs4 import BeautifulSoup
import plotly.graph_objects as go
from plotly.subplots import make_subplots

## Define Graphing Function


In this section, we define the function `make_graph`. You don't have to know how the function works, you should only care about the inputs. It takes a dataframe with stock data (dataframe must contain Date and Close columns), a dataframe with revenue data (dataframe must contain Date and Revenue columns), and the name of the stock.


In [31]:
def make_graph(stock_data, revenue_data, stock):
    fig = make_subplots(rows=2, cols=1, shared_xaxes=True, subplot_titles=("Historical Share Price", "Historical Revenue"), vertical_spacing = .3)
    stock_data_specific = stock_data[stock_data.Date <= '2021--06-14']
    revenue_data_specific = revenue_data[revenue_data.Date <= '2021-04-30']
    fig.add_trace(go.Scatter(x=pd.to_datetime(stock_data_specific.Date, infer_datetime_format=True), y=stock_data_specific.Close.astype("float"), name="Share Price"), row=1, col=1)
    fig.add_trace(go.Scatter(x=pd.to_datetime(revenue_data_specific.Date, infer_datetime_format=True), y=revenue_data_specific.Revenue.astype("float"), name="Revenue"), row=2, col=1)
    fig.update_xaxes(title_text="Date", row=1, col=1)
    fig.update_xaxes(title_text="Date", row=2, col=1)
    fig.update_yaxes(title_text="Price ($US)", row=1, col=1)
    fig.update_yaxes(title_text="Revenue ($US Millions)", row=2, col=1)
    fig.update_layout(showlegend=False,
    height=900,
    title=stock,
    xaxis_rangeslider_visible=True)
    fig.show()

## Question 1: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is Tesla and its ticker symbol is `TSLA`.


In [ ]:
!pip install yfinance
import yfinance as yf
Tesla=yf.Ticker('TSLA')

In [ ]:
Using the ticker object and the function `history` extract stock information and save it in a dataframe named `tesla_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [7]:
tesla_data=df.history(period="max")

NameError: name 'df' is not defined

**Reset the index** using the `reset_index(inplace=True)` function on the tesla_data DataFrame and display the first five rows of the `tesla_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 1 to the results below.


In [ ]:
tesla_data.reset_index(inplace=True)
print(tesla_data.head())

In [20]:
import yfinance as yf
Tesla=yf.Ticker('TSLA')
tesla_data=Tesla.history(period="max")
tesla_data.reset_index(inplace=True)
print(tesla_data.head())

                       Date      Open      High       Low     Close  \
0 2010-06-29 00:00:00-04:00  1.266667  1.666667  1.169333  1.592667   
1 2010-06-30 00:00:00-04:00  1.719333  2.028000  1.553333  1.588667   
2 2010-07-01 00:00:00-04:00  1.666667  1.728000  1.351333  1.464000   
3 2010-07-02 00:00:00-04:00  1.533333  1.540000  1.247333  1.280000   
4 2010-07-06 00:00:00-04:00  1.333333  1.333333  1.055333  1.074000   

      Volume  Dividends  Stock Splits  
0  281494500        0.0           0.0  
1  257806500        0.0           0.0  
2  123282000        0.0           0.0  
3   77097000        0.0           0.0  
4  103003500        0.0           0.0  


## Question 2: Use Webscraping to Extract Tesla Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm Save the text of the response as a variable named `html_data`.


In [8]:
import requests

url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm"
response = requests.get(url)

if response.status_code == 200
    html_data = response.text
else:
    print("Failed to retrieve the webpage. Status code:", response.status_code)


SyntaxError: invalid syntax (2784014090.py, line 6)

Parse the html data using `beautiful_soup`.


In [ ]:
from bs4 import BeautifulSoup
soup = BeautifulSoup(html_data, 'html.parser')


Using `BeautifulSoup` or the `read_html` function extract the table with `Tesla Revenue` and store it into a dataframe named `tesla_revenue`. The dataframe should have columns `Date` and `Revenue`.


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [ ]:
from bs4 import BeautifulSoup
import pandas as pd
# Parse the HTML data using BeautifulSoup

soup = BeautifulSoup(html_data, 'html.parser')
table = soup.find('table', {'class': 'wikitable'})
table_data = []
for row in table.find_all('tr')[1:]:
    columns = row.find_all('td')
    date = columns[0].text.strip()
    revenue = columns[1].text.strip()
    table_data.append({'Date': date, 'Revenue': revenue})
tesla_revenue = pd.DataFrame(table_data)
print(tesla_revenue.head())


Execute the following line to remove the comma and dollar sign from the `Revenue` column. 


In [9]:
tesla_revenue["Revenue"] = tesla_revenue['Revenue'].str.replace(',|\$',"")

NameError: name 'tesla_revenue' is not defined

Execute the following lines to remove an null or empty strings in the Revenue column.


In [ ]:
tesla_revenue.dropna(inplace=True)

tesla_revenue = tesla_revenue[tesla_revenue['Revenue'] != ""]

Display the last 5 row of the `tesla_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [22]:

import requests

url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm"
response = requests.get(url)

if response.status_code == 200:
    html_data = response.text
else:
    print("Failed to retrieve the webpage. Status code:", response.status_code)
from bs4 import BeautifulSoup
import pandas as pd
# Parse the HTML data using BeautifulSoup

soup = BeautifulSoup(html_data, 'html.parser')
table = soup.find('table', {'class': 'wikitable'})
table_data = []
for row in table.find_all('tr')[1:]:
    columns = row.find_all('td')
    date = columns[0].text.strip()
    revenue = columns[1].text.strip()
    table_data.append({'Date': date, 'Revenue': revenue})
tesla_revenue = pd.DataFrame(table_data)
print(tesla_revenue.head())
print(tesla_revenue.tail(5))

AttributeError: 'NoneType' object has no attribute 'find_all'

## Question 3: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is GameStop and its ticker symbol is `GME`.


In [16]:
import yfinance as yf
ticker_gme = yf.Ticker("GME")

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `gme_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [ ]:
gme_data=ticker_gme.history(period="max")

**Reset the index** using the `reset_index(inplace=True)` function on the gme_data DataFrame and display the first five rows of the `gme_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 3 to the results below.


In [ ]:
gme_data.reset_index(inplace=True)
print(gme_data.head())

In [18]:

import yfinance as yf
ticker_gme = yf.Ticker("GME")
gme_data=ticker_gme.history(period="max")
gme_data.reset_index(inplace=True)
print(gme_data.head())

                       Date      Open      High       Low     Close    Volume  \
0 2002-02-13 00:00:00-05:00  1.620128  1.693350  1.603296  1.691667  76216000   
1 2002-02-14 00:00:00-05:00  1.712707  1.716073  1.670626  1.683250  11021600   
2 2002-02-15 00:00:00-05:00  1.683250  1.687458  1.658001  1.674834   8389600   
3 2002-02-19 00:00:00-05:00  1.666417  1.666417  1.578047  1.607504   7410400   
4 2002-02-20 00:00:00-05:00  1.615920  1.662210  1.603296  1.662210   6892800   

   Dividends  Stock Splits  
0        0.0           0.0  
1        0.0           0.0  
2        0.0           0.0  
3        0.0           0.0  
4        0.0           0.0  


## Question 4: Use Webscraping to Extract GME Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html. Save the text of the response as a variable named `html_data`.


In [ ]:
import requests

# URL of the webpage to download
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html"
response = requests.get(url)
if response.status_code == 200:
    html_data = response.text
else:
    print("Failed to retrieve the webpage. Status code:", response.status_code)



Parse the html data using `beautiful_soup`.


In [ ]:
from bs4 import BeautifulSoup
soup = BeautifulSoup(html_data, 'html.parser')


Using `BeautifulSoup` or the `read_html` function extract the table with `GameStop Revenue` and store it into a dataframe named `gme_revenue`. The dataframe should have columns `Date` and `Revenue`. Make sure the comma and dollar sign is removed from the `Revenue` column using a method similar to what you did in Question 2.


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [ ]:
from bs4 import BeautifulSoup
import pandas as pd
# Parse the HTML data using BeautifulSoup
soup = BeautifulSoup(html_data, 'html.parser')
table = soup.find('table', {'id': 'gamestop-revenue'})
table_data = []
for row in table.find_all('tr')[1:]:
    columns = row.find_all('td')
    date = columns[0].text.strip()
    revenue = columns[1].text.strip()
    table_data.append({'Date': date, 'Revenue': revenue})
gme_revenue = pd.DataFrame(table_data)
print(gme_revenue.head())


Display the last five rows of the `gme_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [ ]:
print(gme_revenue.tail(5))

In [29]:
import requests
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html"
response = requests.get(url)
if response.status_code == 200:
    html_data = response.text
else:
    print("Failed to retrieve the webpage. Status code:", response.status_code)
from bs4 import BeautifulSoup
import pandas as pd
# Parse the HTML data using BeautifulSoup
soup = BeautifulSoup(html_data, 'html.parser')
table = soup.find('table', {'id': 'gamestop-revenue'})
for row in table.find_all('tr')[1]:
    columns = row.find_all('td')
    date = columns[0].text.strip()
    revenue = columns[1].text.strip()
    table_data.append({'Date': date, 'Revenue': revenue})
gme_revenue = pd.DataFrame(table_data)
print(gme_revenue.head())
print(gme_revenue.tail(5))


AttributeError: 'NoneType' object has no attribute 'find_all'

## Question 5: Plot Tesla Stock Graph


Use the `make_graph` function to graph the Tesla Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(tesla_data, tesla_revenue, 'Tesla')`. Note the graph will only show data upto June 2021.


In [61]:


# Filter data up to June 2021
tesla_data = tesla_data[tesla_data['Date'] <= '2021-06-30']
tesla_revenue = tesla_revenue[tesla_revenue['Date'] <= '2021-06-30']

# Define your custom make_graph function (replace with your actual function)
def make_graph(tesla_data, tesla_revenue,'Tesla'):
    # Your custom graphing logic here using Plotly
    # Create Plotly figures, add traces, customize layout, etc.
    # Example:
    import plotly.graph_objs as go
    import plotly.subplots as sp

    # Create a subplot with two vertical subplots
    fig = sp.make_subplots(rows=2, cols=1, shared_xaxes=True, vertical_spacing=0.02)

    # Add a trace for Tesla stock data
    trace_stock = go.Scatter(x=stock_data['Date'], y=stock_data['Close'], mode='lines', name='Tesla Stock Price')
    fig.add_trace(trace_stock, row=1, col=1)

    # Add a trace for Tesla revenue data (assuming revenue has 'Date' and 'Revenue' columns)
    trace_revenue = go.Scatter(x=revenue_data['Date'], y=revenue_data['Revenue'], mode='lines', name='Tesla Revenue')
    fig.add_trace(trace_revenue, row=2, col=1)

    # Customize the layout
    fig.update_layout(
        title=title + ' Stock and Revenue Data up to June 2021',
        xaxis=dict(title='Date'),
        yaxis=dict(title='Stock Price (USD)', row=1, col=1),
        yaxis=dict(title='Revenue', row=2, col=1)
    )

    # Show the graph
    fig.show()

# Call your custom make_graph function
make_graph(tesla_data, tesla_revenue, 'Tesla')


SyntaxError: invalid syntax (2793562630.py, line 6)

## Question 6: Plot GameStop Stock Graph


Use the `make_graph` function to graph the GameStop Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(gme_data, gme_revenue, 'GameStop')`. Note the graph will only show data upto June 2021.


In [62]:
gme_data = gme_data[gme_data['Date'] <= '2021-06-30']
gme_revenue = gme_revenue[gme_revenue['Date'] <= '2021-06-30']

# Define your custom make_graph function (replace with your actual function)
def make_graph(gme_data, gme_revenue,'GameStop'):
    # Your custom graphing logic here using Plotly
    # Create Plotly figures, add traces, customize layout, etc.
    # Example:
    import plotly.graph_objs as go
    import plotly.subplots as sp
    # Create a subplot with two vertical subplots
    fig = sp.make_subplots(rows=2, cols=1, shared_xaxes=True, vertical_spacing=0.02)
    # Add a trace for GameStop stock data
    trace_stock = go.Scatter(x=stock_data['Date'], y=stock_data['Close'], mode='lines', name='GameStop Stock Price')
    fig.add_trace(trace_stock, row=1, col=1)

    # Add a trace for GameStop revenue data (assuming revenue has 'Date' and 'Revenue' columns)
    trace_revenue = go.Scatter(x=revenue_data['Date'], y=revenue_data['Revenue'], mode='lines', name='GameStop Revenue')
    fig.add_trace(trace_revenue, row=2, col=1)

    # Customize the layout
    fig.update_layout(
        title=title + ' Stock and Revenue Data up to June 2021',
        xaxis=dict(title='Date'),
        yaxis=dict(title='Stock Price (USD)', row=1, col=1),
        yaxis=dict(title='Revenue', row=2, col=1)
    )
    # Show the graph
    fig.show()
# Call your custom make_graph function
make_graph(gme_data, gme_revenue, 'GameStop')


SyntaxError: keyword argument repeated (3732856938.py, line 28)

<h2>About the Authors:</h2> 

<a href="https://www.linkedin.com/in/joseph-s-50398b136/">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.

Azim Hirjani


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By    | Change Description        |
| ----------------- | ------- | ------------- | ------------------------- |
| 2022-02-28        | 1.2     | Lakshmi Holla | Changed the URL of GameStop |
| 2020-11-10        | 1.1     | Malika Singla | Deleted the Optional part |
| 2020-08-27        | 1.0     | Malika Singla | Added lab to GitLab       |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>

<p>
